### PR TITLE
Signal interface and derived classes

### LINKED ISSUE #1038: Add Signal classes to encapsulate MSR to double conversion logic
> estimate 5

Add a Signal interface that can be used for raw MSRs, MSR fields, and signals that are a combination of other signals.

Parent story #435

### PR DESCRIPTION
**Pull request by dianarg**
Adds classes that will be used to support MSRIOGroup refactoring.  These classes will replace MSRSignal, CombinedSignal & friends, and parts of the MSR + MSREncode usage in MSRIOGroup.

Resolves #1038

### CODE REVIEW

File: src/RawMSRSignal.cpp [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    {
+        return geopm::make_unique<RawMSRSignal>(*this);
+    }
+
+    void RawMSRSignal::setup_batch(void)
+    {
+        GEOPM_DEBUG_ASSERT(m_msrio != nullptr, "no valid MSRIO object.");
+
+        if (!m_is_batch_ready) {
+            m_data = m_msrio->add_read(m_cpu, m_offset);
```
Comment (dianarg) [inline src/RawMSRSignal.cpp:61]: One change we discussed was to have this data pointer not be exposed here and add a sample() method to MSRIO.

File: src/DerivativeSignal.cpp [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+        : m_time_sig(time_sig)
+        , m_y_sig(y_sig)
+        , M_NUM_SAMPLE_HISTORY(num_sample_history)
+        , m_history(M_NUM_SAMPLE_HISTORY)
+        , m_derivative_num_fit(0)
+        , m_is_batch_ready(false)
+        , m_sleep_time(sleep_time)
+    {
+        GEOPM_DEBUG_ASSERT(m_time_sig && m_y_sig,
+                           "underlying Signals cannot be null.");
```
Comment (bgeltz) [inline src/DerivativeSignal.cpp]: Underlying signals...

File: src/DerivativeSignal.hpp [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+namespace geopm
+{
+    class DerivativeSignal : public Signal
+    {
+        public:
+            DerivativeSignal(std::shared_ptr<Signal> time_sig,
+                             std::shared_ptr<Signal> y_sig,
+                             int read_loops, double sleep_time);
+            DerivativeSignal(const DerivativeSignal &other);
```
Comment (bgeltz) [inline src/DerivativeSignal.hpp]: Is the copy constructor used anywhere?

File: src/DerivativeSignal.cpp [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    {
+        double result = NAN;
+        CircularBuffer<m_sample_s> temp_history(M_NUM_SAMPLE_HISTORY);
+        int num_fit = 0;
+        for (int ii = 0; ii < M_NUM_SAMPLE_HISTORY; ++ii) {
+            double time = m_time_sig->read();
+            double signal = m_y_sig->read();
+            result = compute_next(temp_history, num_fit, time, signal);
+            if (ii < M_NUM_SAMPLE_HISTORY) {
+                usleep(m_sleep_time);            }
```
Comment (bgeltz) [inline src/DerivativeSignal.cpp]: Closing curly is in the wrong spot.

File: src/DifferenceSignal.cpp [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+namespace geopm
+{
+    DifferenceSignal::DifferenceSignal(std::shared_ptr<Signal> minuend,
+                                       std::shared_ptr<Signal> subtrahend)
+        : m_minuend(minuend)
+        , m_subtrahend(subtrahend)
+        , m_is_batch_ready(false)
+    {
+        GEOPM_DEBUG_ASSERT(m_minuend && m_subtrahend,
+                           "underlying Signals cannot be null.");
```
Comment (bgeltz) [inline src/DifferenceSignal.cpp]: Same capitalization issues as before

File: src/DifferenceSignal.hpp [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+namespace geopm
+{
+    /// A composite signal used by an IOGroup to produce a signal as
+    /// the difference of two signals.
+    class DifferenceSignal : public Signal
+    {
+        public:
+            DifferenceSignal(std::shared_ptr<Signal> minuend,
+                             std::shared_ptr<Signal> subtrahend);
+            DifferenceSignal(const DifferenceSignal &other);
```
Comment (bgeltz) [inline src/DifferenceSignal.hpp]: Is the copy constructor used?

File: src/MSRFieldSignal.hpp [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    /// MSREncode.  The enum for the function comes from the MSR class.
+    class MSRFieldSignal : public Signal
+    {
+        public:
+            MSRFieldSignal(std::shared_ptr<Signal> raw_msr,
+                           int begin_bit,
+                           int end_bit,
+                           int function,
+                           double scalar);
+            MSRFieldSignal(const MSRFieldSignal &other);
```
Comment (bgeltz) [inline src/MSRFieldSignal.hpp]: ?

**Review by bgeltz — CHANGES_REQUESTED** [review]
Looks mostly good to me.

A few nits on capitalization and curly brace placement.

Do we need all those copy constructors for the various signal types?  I didn't see where they were used.

File: src/DerivativeSignal.cpp (hunk 2 above)
Comment (dianarg) [reply src/DerivativeSignal.cpp]: Maybe something like "Signal pointers for time_sig and y_sig cannot be null" ?

File: src/DerivativeSignal.hpp (hunk 3 above)
Comment (dianarg) [reply src/DerivativeSignal.hpp]: It's used in clone()

**Review by dianarg — COMMENTED** [review]
The copy constructors are needed because when the MSRIOGroup implements read_signal(), it will need to copy a shared_ptr<Signal> (case class pointer) but we need to know which copy constructor to call.  Also, some of the Signal types have state so the copy constructors are non-trivial.

**Review by bgeltz — APPROVED** [review]

**Pull request merged**